### ISSUE TITLE
How to save trained SimCLR model and load it again to get test data embeddings

### ISSUE THREAD

**Issue by Ashufet** [lightly-ai/lightly#287]
new_collate_fn = lightly.data.SimCLRCollateFunction(
    input_size=input_size,
    vf_prob=0.5,
    rr_prob=0.5,
    cj_prob=0.0,
    random_gray_scale=0.0
)

# let's update our collate method and reuse our dataloader
dataloader_train_simclr.collate_fn=new_collate_fn

# create a ResNet backbone and remove the classification head
resnet = lightly.models.ResNetGenerator('resnet-18')
last_conv_channels = list(resnet.children())[-1].in_features
backbone = nn.Sequential(
    *list(resnet.children())[:-1],
    nn.Conv2d(last_conv_channels, num_ftrs, 1),
    nn.AdaptiveAvgPool2d(1)
)
model = lightly.models.SimCLR(backbone, num_ftrs=num_ftrs)
optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
encoder = lightly.embedding.SelfSupervisedEmbedding(
    model,
    criterion,
    optimizer,
    dataloader_train_simclr
)

encoder.train_embedding(gpus=gpus,
                        progress_bar_refresh_rate=100,
                        max_epochs=max_epochs)
encoder = encoder.to(device)

embeddings, _, fnames = encoder.embed(dataloader_test, device=device)
embeddings = normalize(embeddings)

**I want to save the encoder and load it again to do the inference. How can I do it?**

**Comment by IgorSusmelj** [comment]
Hi @Ashufet,

Something like this? I'm currently updating the documentation to make this more clear. 

```python
resnet = lightly.models.ResNetGenerator('resnet-18')
last_conv_channels = list(resnet.children())[-1].in_features
backbone = nn.Sequential(
    *list(resnet.children())[:-1],
    nn.Conv2d(last_conv_channels, num_ftrs, 1),
    nn.AdaptiveAvgPool2d(1)
)
model = lightly.models.SimCLR(backbone, num_ftrs=num_ftrs)

# you can always access the backbone model using `.backbone`
# model.backbone

# you can also store the backbone and use it in another code
state_dict = {
    'resnet18_parameters': model.backbone.state_dict()
}
torch.save(state_dict, 'model.pth')


# ------- THIS COULD BE IN A NEW FILE (e.g. inference.py) -------
# Make sure you place the `model.pth` file in the same folder as this code

# load the model in a new file for inference
resnet_new = lightly.models.ResNetGenerator('resnet-18')

# note that we need to create exactly the same backbone in order to load the weights
last_conv_channels = list(resnet_new.children())[-1].in_features
backbone_new = nn.Sequential(
    *list(resnet_new.children())[:-1],
    nn.Conv2d(last_conv_channels, num_ftrs, 1),
    nn.AdaptiveAvgPool2d(1)
)

ckpt = torch.load('model.pth')
backbone_new.load_state_dict(ckpt['resnet18_parameters'])

# for inference of the resnet (encoder) you can switch it to eval mode and pass it a tensor
backbone_new.eval()
output = backbone_new(some_input_tensor)
```

Please let me know whether this helps you!

**Comment by Ashufet** [comment]
Thanks @IgorSusmelj , It's working.

**Comment by Samjith888** [comment]
Thanks for the great work. I have trained a model by using SimCLR. Tried to run the inference on a image.


```
img = Image.open("/home/usr/images/test/belts/46.JPEG")
tran = transforms.ToTensor() 
img_tr = tran(img)
output = backbone_new(img_tr[None, ...])
```

How to get  the similar images of the input image from this ?

**Comment by philippmwirth** [comment]
Dear @Samjith888 

To find similar images you will need to embed all images (query and keys) and then do a nearest neighbor search.

Our [tutorial on SimCLR](https://docs.lightly.ai/tutorials/package/tutorial_simclr_clothing.html) has sample code on how to do this.

**Comment by Samjith888** [comment]
> Dear @Samjith888
> 
> To find similar images you will need to embed all images (query and keys) and then do a nearest neighbor search.
> 
> Our [tutorial on SimCLR](https://docs.lightly.ai/tutorials/package/tutorial_simclr_clothing.html) has sample code on how to do this.

I have trained a model (model.pth) by using a dataset . Can i get embeddings of the trained  images from that trained weight ?

**Comment by philippmwirth** [comment]
Yes, simply load the model from the checkpoint, and get the embeddings by passing the images you are interested in through the backbone of the pre-trained model.

Let me know if that helps you :)

**Comment by Samjith888** [comment]
So if i want to get all embedddings of train images from a trained checkpoint , then i should  pass the train_dataset_loader into the below function right ?

`output = backbone_new(train_dataset_loader)`
